### PR TITLE
Integer Only Common Operations

### LINKED ISSUE #100: Virtual Machine - Implement Integer Only Common Operations
Integer only common operations only apply on integers i.e. `i32` and `i64`. These are:
- [x] 0x39 - And
- [x] 0x3a - Or
- [x] 0x3b - Xor
- [x] 0x3c - Shl
- [x] 0x3d - ShrSigned
- [x] 0x3e - ShrUnsigned
- [x] 0x3f - Rotl
- [ ] 0x40 - Rotr

### PR DESCRIPTION
**Pull request by iNemesis21**
Hello.
This PR closes #100

### CODE REVIEW

File: src/purple_vm/src/primitives/value.rs [hunk 1]
```diff
@@ -612,6 +612,154 @@ impl VmValue {
...
+            }
+            (VmValue::f32Array2(_), VmValue::f32Array2(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f32Array4(_), VmValue::f32Array4(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f32Array8(_), VmValue::f32Array8(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array2(_), VmValue::f64Array2(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array4(_), VmValue::f64Array4(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array8(_), VmValue::f64Array8(_)) => Err(VmError::InvalidOperator),
+            (_, _) => panic!("Cannot perform rotate_left between different variants!"),
+        }
+    }
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Could you please add match clauses for arrays up to 256 items?

File: src/purple_vm/src/primitives/value.rs [hunk 2]
```diff
@@ -612,6 +612,154 @@ impl VmValue {
...
+            }
+            (VmValue::f32Array2(_), VmValue::f32Array2(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f32Array4(_), VmValue::f32Array4(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f32Array8(_), VmValue::f32Array8(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array2(_), VmValue::f64Array2(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array4(_), VmValue::f64Array4(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array8(_), VmValue::f64Array8(_)) => Err(VmError::InvalidOperator),
+            (_, _) => panic!("Cannot perform rotate_right between different variants!"),
+        }
+    }
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please add match clauses for arrays up to 256 items in length.

File: src/purple_vm/src/primitives/value.rs [hunk 3]
```diff
@@ -2190,6 +2338,508 @@ impl Neg for VmValue {
...
+            (VmValue::f32Array2(_), VmValue::f32Array2(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f32Array4(_), VmValue::f32Array4(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f32Array8(_), VmValue::f32Array8(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array2(_), VmValue::f64Array2(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array4(_), VmValue::f64Array4(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array8(_), VmValue::f64Array8(_)) => Err(VmError::InvalidOperator),
+            (_, _) => panic!("Cannot perform bitand between different variants!"),
+        }
+    }
+}
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please add match clauses for arrays up to 256 items in length

File: src/purple_vm/src/primitives/value.rs [hunk 4]
```diff
@@ -2190,6 +2338,508 @@ impl Neg for VmValue {
...
+            (VmValue::f32Array2(_), VmValue::f32Array2(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f32Array4(_), VmValue::f32Array4(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f32Array8(_), VmValue::f32Array8(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array2(_), VmValue::f64Array2(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array4(_), VmValue::f64Array4(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array8(_), VmValue::f64Array8(_)) => Err(VmError::InvalidOperator),
+            (_, _) => panic!("Cannot perform bitor between different variants!"),
+        }
+    }
+}
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please add match clauses for arrays up to 256 items in length

File: src/purple_vm/src/primitives/value.rs [hunk 5]
```diff
@@ -2190,6 +2338,508 @@ impl Neg for VmValue {
...
+            (VmValue::f32Array2(_), VmValue::f32Array2(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f32Array4(_), VmValue::f32Array4(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f32Array8(_), VmValue::f32Array8(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array2(_), VmValue::f64Array2(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array4(_), VmValue::f64Array4(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array8(_), VmValue::f64Array8(_)) => Err(VmError::InvalidOperator),
+            (_, _) => panic!("Cannot perform bitxor between different variants!"),
+        }
+    }
+}
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please add match clauses for arrays up to 256 items in length.

File: src/purple_vm/src/primitives/value.rs [hunk 6]
```diff
@@ -2190,6 +2338,508 @@ impl Neg for VmValue {
...
+            (VmValue::f32Array2(_), VmValue::f32Array2(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f32Array4(_), VmValue::f32Array4(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f32Array8(_), VmValue::f32Array8(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array2(_), VmValue::f64Array2(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array4(_), VmValue::f64Array4(_)) => Err(VmError::InvalidOperator),
+            (VmValue::f64Array8(_), VmValue::f64Array8(_)) => Err(VmError::InvalidOperator),
+            (_, _) => panic!("Cannot perform shr between different variants!"),
+        }
+    }
+}
```
Comment (octavonce) [inline src/purple_vm/src/primitives/value.rs]: Please add match clauses for arrays up to 256 items in length.

**Review by octavonce — CHANGES_REQUESTED** [review]
Hi there!

This looks great, with only a few pointers left in the review.

**Comment by iNemesis21** [comment]
Hi @octavonce,
I've finished the implementation for this PR and I've handled issue #143 also.
Since the const-generics feature of rust is not stable yet, for arrays larger than 32 I've made conversions to vecs, but I left a TODO so when the feature is available and stable, we should get rid of the conversions.

**Comment by octavonce** [comment]
@iNemesis21 That's good enough for now. I'm going to merge this as is. Could you also please open an issue about getting rid of `to_vec()`s so it will be documented?

**Review by octavonce — APPROVED** [review]

**Pull request merged**